src/common/io_exerciser/DataGenerator.h

```diff
@@ -0,0 +1,194 @@
+#pragma once
+
+#include <memory>
+#include <random>
+
+#include "include/buffer.h"
+#include "ObjectModel.h"
+
+/* Overview
+ *
+ * class DataGenerator
+ *   Generates data buffers for write I/Os using state queried
+ *   from ObjectModel. Validates data buffers for read I/Os
+ *   against the state in the ObjectModel. If a data miscompare
+ *   is detected provide debug information about the state of the
+ *   object, the buffer that was read and the expected buffer.
+ *
+ *
+ * class DataGenerator::DataGenerationSingleton
+ *   A singleton object created on first use to create and store
+ *   a unique run identifier. Can either be created supplying an id
+ *   or it can be generated at first use. Id cannot and should not be
+ *   changed after being created.
+ *
+ * 
+ * class SeededRandomGenerator
+ *   Inherits from DataGenerator. Generates entirely random patterns
+ *   based on the seed retrieved by the model.
+ * 
+ * 
+ * class HeaderedSeededRandomGenerator
+ *   Inherits from SeededDataGenerator. Generates entirely random patterns
+ *   based on the seed retrieved by the model, however also appends a 
+ *   header to the start of each block. This generator also provides
+ *   a range of verbose debug options to help disagnose a miscompare
+ *   whenever it detects unexpected data.
+ */
+ 
+namespace ceph {
+  namespace io_exerciser {
+    namespace data_generation {
+      enum class GenerationType {
+        SeededRandom,
+        HeaderedSeededRandom
+        // CompressedGenerator
+        // MixedGenerator
+      };
+    
+      class DataGenerator {
+      public:
+        virtual ~DataGenerator() = default;
+        static std::unique_ptr<DataGenerator>
+          create_generator(GenerationType generatorType,
+                           const ObjectModel& model);
+        virtual void generate_data(uint64_t length, uint64_t offset,
+                                   bufferlist& retlist)=0;
+        virtual bool validate(bufferlist& bufferlist, uint64_t offset,
+                              uint64_t length);
+
+        // Used for testing debug outputs from data generation
+        virtual void generate_wrong_data(uint64_t offset, uint64_t length,
+                                         bufferlist& retlist);
+
+      protected:
+        const ObjectModel& m_model;
+
+        DataGenerator(const ObjectModel& model) : m_model(model) {}
+
+        class DataGenerationSingleton {
+        private:
+          DataGenerationSingleton();
+          DataGenerationSingleton(uint64_t unique_id);
+
+          bool m_created = false;
+          uint64_t m_uniqueId = 0;
+
+          static DataGenerationSingleton m_singletonInstance;
+          static std::mutex m_mutex;
+
+        public:
+          static const DataGenerationSingleton&
+            createSpecificInstance(uint64_t unique_id);
+          static const DataGenerationSingleton& getInstance();
+          static const uint64_t getUniqueId();
```

Comment:
no point in the 'const' on a returned uint. 
More important: note that the decision to use a singleton for the DataGenerator
object means, for example, that you have at least two more static methods.
Can the DG instance be injected to its users, instead of using a singleton?

Reply:
Have refactored out the use of the singleton and instead generate the unique ID optionally with the HeaderedSeededRandomGenerator, as this was the only class that I intended to make use of this